src/coreclr/md/enc/sha256.cpp

```diff
@@ -0,0 +1,141 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+//*****************************************************************************
+// sha256.cpp
+//
+
+//
+// contains implementation of sha256 hash algorithm
+//
+//*****************************************************************************
+
+#ifdef _WIN32
+HRESULT Sha256Hash(BYTE* pSrc, DWORD srcSize, BYTE* pDst, DWORD dstSize)
+{
+    NTSTATUS status;
+
+    BCRYPT_ALG_HANDLE   algHandle = NULL;
+    BCRYPT_HASH_HANDLE  hashHandle = NULL;
+
+    BYTE    hash[32]; // 256 bits
+    DWORD   hashLength = 0;
+    DWORD   resultLength = 0;
+    status = BCryptOpenAlgorithmProvider(&algHandle, BCRYPT_SHA256_ALGORITHM, NULL, BCRYPT_HASH_REUSABLE_FLAG);
+    if(!NT_SUCCESS(status))
+    {
+        goto cleanup;
+    }
+    status = BCryptGetProperty(algHandle, BCRYPT_HASH_LENGTH, (PBYTE)&hashLength, sizeof(hashLength), &resultLength, 0);
+    if(!NT_SUCCESS(status))
+    {
+        goto cleanup;
+    }
+    if (hashLength != 32)
+    {
+        status = STATUS_NO_MEMORY;
+        goto cleanup;
+    }
+    status = BCryptCreateHash(algHandle, &hashHandle, NULL, 0, NULL, 0, 0);
+    if(!NT_SUCCESS(status))
+    {
+        goto cleanup;
+    }
+
+    status = BCryptHashData(hashHandle, pSrc, srcSize, 0);
+    if(!NT_SUCCESS(status))
+    {
+        goto cleanup;
+    }
+
+    status = BCryptFinishHash(hashHandle, hash, hashLength, 0);
+    if(!NT_SUCCESS(status))
+    {
+        goto cleanup;
+    }
+
+    memcpy(pDst, hash, min(hashLength, dstSize));
+    status = S_OK;
+
+cleanup:
+    if (NULL != hashHandle)
+    {
+         BCryptDestroyHash(hashHandle);
+    }
+    if(NULL != algHandle)
+    {
+        BCryptCloseAlgorithmProvider(algHandle, 0);
+    }
+    return status;
+}
+#elif defined(__APPLE__)
+#include <CommonCrypto/CommonCrypto.h>
+#include <CommonCrypto/CommonDigest.h>
+
+HRESULT Sha256Hash(BYTE* pSrc, DWORD srcSize, BYTE* pDst, DWORD dstSize)
+{
+    CC_SHA256_CTX ctx = {{ 0 }};
+
+    if (!CC_SHA256_Init(&ctx))
```

Comment:
Apple has a one-shot you can use too. It would look something like this:

```C
 HRESULT Sha256Hash(BYTE* pSrc, DWORD srcSize, BYTE* pDst, DWORD dstSize)
 {
     if (dstSize < CC_SHA256_DIGEST_LENGTH)
     {
         return E_FAIL;
     }
         
     CC_SHA256(pSrc, (CC_LONG)srcSize, pDst);
     return S_OK;
 }
```

If there is an actual need for it, you can copy to a temp buffer like Windows and Linux do.